joint_state_publisher/joint_state_publisher/joint_state_publisher.py

```diff
@@ -99,36 +107,60 @@ def init_sdf(self, robot):
 
             if jtype == 'continuous':
                 joint['continuous'] = True
-            self.free_joints[name] = joint
-            self.joint_list.append(name)
+            free_joints[name] = joint
+            joint_list.append(name)
 
-    def init_collada(self, robot):
-        robot = robot.getElementsByTagName('kinematics_model')[0].getElementsByTagName('technique_common')[0]
-        for child in robot.childNodes:
+        return (free_joints, joint_list, dependent_joints)
+
+    def init_collada(self, xmldom):
+        free_joints = {}
+        joint_list = []
+        dependent_joints = {}
+
+        colladadom = xmldom.childNodes[0]
+
+        if not colladadom.hasAttribute('version'):
+            raise Exception('COLLADA file must have a version tag')
+
+        colladaversion = packaging.version.parse(colladadom.attributes['version'].value)
+        if colladaversion < packaging.version.parse('1.5.0'):
+            raise Exception('COLLADA file must be at least version 1.5.0')
+
+        kinematics_model = xmldom.getElementsByTagName('kinematics_model')[0]
+        technique_common = kinematics_model.getElementsByTagName('technique_common')[0]
+        for child in technique_common.childNodes:
             if child.nodeType is child.TEXT_NODE:
                 continue
-            if child.localName == 'joint':
-                name = child.getAttribute('name')
-                if child.getElementsByTagName('revolute'):
-                    joint = child.getElementsByTagName('revolute')[0]
-                else:
-                    self.get_logger().warn('Unknown joint type %s', child)
-                    continue
+            if child.localName != 'joint':
+                continue
 
-                if joint:
-                    limit = joint.getElementsByTagName('limits')[0]
-                    minval = float(limit.getElementsByTagName('min')[0].childNodes[0].nodeValue)
-                    maxval = float(limit.getElementsByTagName('max')[0].childNodes[0].nodeValue)
-                if minval == maxval:  # this is a fixed joint
-                    continue
+            name = child.getAttribute('name')
+            if child.getElementsByTagName('revolute'):
+                joint = child.getElementsByTagName('revolute')[0]
+            else:
+                self.get_logger().warn('Unknown joint type %s', child)
+                continue
+
+            if joint:
+                limit = joint.getElementsByTagName('limits')[0]
+                minval = float(limit.getElementsByTagName('min')[0].childNodes[0].nodeValue)
+                maxval = float(limit.getElementsByTagName('max')[0].childNodes[0].nodeValue)
```

Comment:
Should we catch `ValueError` or limits not specified perhaps?

Reply:
This is another good point.

Right now, we do different things between `init_sdf`/`init_urdf` and `init_collada`.  In the former case, we actually catch the exception of a missing limit, print a warning message, skip the current joint, and continue processing the rest of the joints.  In the latter case, we currently fail the entire process, and don't update the XML at all.

I think the `init_collada` way is actually better.  That is, I think it is a lot more noticeable to users if we fail to parse the *entire* description, rather than just be missing one joint.  That latter behavior can be easy to miss and hard to diagnose.  I'm actually going to update this PR to remove the special handling of this in `init_sdf` and `init_urdf`.

Reply:
Agreed but even if we don't want to ignore the joint and move forward, I think there could be warning messages printed in all three functions as to which joint is lacking limits so it's easier to debug and update the XML.

Reply:
Yes, good point.

After looking at this for a while, I ended up getting myself quite confused.  I think the problem is that we don't have good tests which define what the various parsers do at the moment, so there is no clear way forward on where we want to go.

So I decided to take a step back and write a series of parsing tests.  These tests show what the *current* behavior of the various parsers we have are.  I'll open that as a separate PR which I think we should get in first.  Once we have that, then this PR can make modifications to both the code and the tests, and we can be a lot more confident in the direction we've taken.

Reply:
And see #97 now.